### ISSUE TITLE
Packages unavailable for R version 3.4.2

### ISSUE THREAD

**Issue by Althalis** [DorresteinLaboratory/Kendrick_Mass_Filter#2]
Hi, 
I was trying to use both the ParameterExploration_Final.ipynb and KendrickMassFilter_Final.ipynb, but it seems like some dependencies that you cite and that should be loaded are not available for R 3.4.2 (I actually installed this older version just for that because the same dependencies are not available for R3.4.3 either).
It's the case for Rgraphviz and doMC for instance. And probably more but I haven't tried everything yet since I have the feeling something is wrong...
Any clue?
Maybe I don't need these really, but I'd rather figure what's wrong before spending time. :) :)
See screenshots attached
![billede](https://user-images.githubusercontent.com/5079495/61729493-a1df7300-ad77-11e9-8543-f2633bfed461.png)
![billede](https://user-images.githubusercontent.com/5079495/61729535-b91e6080-ad77-11e9-8848-a7e7b7562645.png)

Thank you!
Julie

**Comment by rsilvabioinfo** [comment]
Hello Julie,

Rgraphviz is not on CRAN, here is how you can install it
https://www.bioconductor.org/packages/release/bioc/html/Rgraphviz.html

Sometimes is hard to manage package versions in R. You can also google package names
https://cran.r-project.org/web/packages/doMC/index.html

And you can download the package and do:
install.packages('PACKAGE_I_DOWNLOADED_FOR_MY_SYSTEM', dependencies=TRUE)

Let us know if have more problems to get it working.

Cheers,
Ricardo

**Comment by Althalis** [comment]
Hello Ricardo, 
thanks for the quick reply to my beginner questions. :)
Do you know if all your scripts could run on R3.6.1 then potentially?
Thanks
Cheers
Julie

**Comment by rsilvabioinfo** [comment]
Hi Julie,

yes, it should work on old versions of R. We are sorry it is not very
friendly to install and use. I will try to come up with some
instructions to help beginners to install packages.

Once you have all package dependencies it should work.
Take a look at this stackoverflow step by step answer on how to check
install errors in R:
https://stackoverflow.com/questions/25721884/how-should-i-deal-with-package-xxx-is-not-available-for-r-version-x-y-z-wa

Thanks,
Ricardo

Em qua, 24 de jul de 2019 às 06:13, Julie Courraud
<notifications@github.com> escreveu:

**Comment by Althalis** [comment]
Hello again, 
So I managed to run everything with a R3.6.1 kernel. Yepee!

doMC: my understanding is that doMC is on CRAN, but if you're using Windows, you should type
_install.packages("doMC", repos="http://R-Forge.R-project.org")_

2 questions:
- I was wondering how to explain that my composition analysis detects around 770 features in the PEG category, but when I run the filter for PEG, I end up with 3700 features filtered.

- The composition analysis also shows PPG (440 features). Could it be a good idea to run the PEG-filtered  matrix into the filter again, but choosing "polymer = "polypropylene_glycol_other_C3H6O1"" instead? Should I adjust other parameters then? 

Thanks! very helpful work :)

**Comment by alan-jarmusch** [comment]
Hi Julie,
I would be happy to help with your questions and any subsequent questions.

**Q1: I was wondering how to explain that my composition analysis detects around 770 features in the PEG category, but when I run the filter for PEG, I end up with 3700 features filtered.**

Re Q1: The difference between the composition analysis is likely the cause of different settings. The composition analysis runs the filter for each potential polymer independently using a set of defaults (from the code: mass_defect_parameter = 0.01, retention_time = 0.8, connection_filter = TRUE, connections_min_number = 5). The default for connections_min_number (or NOS) is 5 which attempts to be more specific at the cost of reduced sensitivity. I would bet that the filter for PEG is using a lower value for NOS and therefore more features are highlighted as potentially being PEG. 

**Q2: The composition analysis also shows PPG (440 features). Could it be a good idea to run the PEG-filtered matrix into the filter again, but choosing "polymer = "polypropylene_glycol_other_C3H6O1"" instead? Should I adjust other parameters then?**

Re Q2: Good question. The compositional analysis is intended to be used in an exploratory fashion as the settings for the filtering (same values used in both compositional analysis and the filter) will change results. We have not done extensive testing of which settings are optimal for each polymer, only PEG, and only in our data set. I suspect the optimal settings are data set specific. I would suggest confirming the findings manually via interpretation of the mass spectra - I'm happy to assist. As for repeated filtering steps for multiple polymers, we have not characterized the performance, but it should work.

Thanks for your interest - feedback is always welcome.
-Alan

**Comment by Althalis** [comment]
Hi Allan, Hi Ricardo, 
sorry for the late reply, I'm trying different things.
Thanks for the clear answers.

I'm not experienced enough to understand the details of your code, so I have a doubt regarding how to adjust the parameters.
(I don't have the necessary data to run the Parameter Exploration but I could try a few combinations)

There are 2 places in the code where KMD, RT, NOS can be changed, as well as the polymer potentially (line 5 and 9 see screenshot below).
1- If I want to filter PEG with a different NOS, say 5, should I adapt NOS value in both places? (line 5 and 9)
2- I guess the second line with KMD=0.95 has a specific purpose. If there's an easy way to explain, I'd be curious to understand (but there is so much I don't understand that it might be a waste of time trying to explain to me haha)
3- Basically, I guess I should not change this KMD=0.95 value in line 9 when I'm testing different parameters. 
4- If I change to PPG, I have the feeling I should adjust the FractionBase on line 9. How do I do that? Which number to use? (this text don't make sense to me, sorry : 
" # Re-scale according to fraction base, for this case
    # ~ fractionBase*mass_defect_parameter*2"

Thanks a lot for clarifying, 
Best
Julie

![billede](https://user-images.githubusercontent.com/5079495/61947199-e865e500-afa4-11e9-838d-83091bb6e67a.png)
![billede](https://user-images.githubusercontent.com/5079495/61947218-f0be2000-afa4-11e9-81fc-e25584dd40eb.png)

**Comment by Althalis** [comment]
Actually I tried with NOS 5 at both lines, and I get 3500 features filtered. So much more than the composition analysis still. Any clue?
Thanks!

**Comment by rsilvabioinfo** [comment]
Hello Julie,

I will attempt to answer the first question, Alan knows the chemistry and
will be able to help with the others.

If you look at the source
<https://github.com/DorresteinLaboratory/Kendrick_Mass_Filter/blob/master/kendrick.mass.filter_new_all_datamatrix_terminologyupdate.R>,
you will see that for polymer choice you can either choose the 'polymer'
parameter or 'fractionBase'. If you choose the  'fractionBase' parameter
the polymer one is ignored. I'm not sure what you mean by multiple places
in the code where you can change KMD and NOS. Above you showed examples of
the function 'Kendrick.mass.filter'. The function is repeated in the
notebook just to illustrate how one can change the parameters.

Alan will help you as soon as possible.

Cheers,
Ricardo



Em sex, 26 de jul de 2019 às 08:34, Julie Courraud <notifications@github.com>
escreveu:

**Comment by Althalis** [comment]
Hi Ricardo, 
Ok, that's what I suspected initially, that this function should be run only once, but it's not very clear that it's just another example (I'll add some text to it). If you're a beginner like me, you just assume all lines are necessary to get to the results, so only the second set of parameters (block 9-10-11) would be used since it probably overwrites the first set (block 5-7-8). Which explains why I have so many filtered features (KMD 0.95) I guess.

So my understanding now is that I should ignore 9-10-11 (the 3 blocks after the formula explanation) which are indeed copies of 5-7-8.

But why is there this piece of code in line 9?
    connection_filter = TRUE,
    fractionBase=43)

This is why I thought there was something special about this part that I shouldn't remove. :)
I'm very sorry that I can't understand your R source and figure that by myself.

So if I want to run PPG for instance, I should run
![billede](https://user-images.githubusercontent.com/5079495/61951132-5532ac80-afb0-11e9-92f8-08e423c7b67b.png)

(If I keep the piece of code with the fractionBase=43, it fails. )

Thanks again, (you've actually addressed all my questions in the end, since they were all connected to this misunderstanding of repeated code).
Best
Julie

**Comment by rsilvabioinfo** [comment]
Hello Julie,

I'm glad I could help with something.

You are right, it is still confusing, lets maybe see your feedback on how
useful this was for you and we can improve from that. For instance, our
conversation is public, and if a new user googles an error message they may
see that, so, congratulations to your first steps on github and coding,
keep trying and you will soon understand it much better.

About your error, you can also post the error message as a separate issue
and we can check whenever we have time.

The KMF has two filtering steps, first, it filters the pairs that match the
KMD provided and then it only keeps the filters for multimer series that
have at least NOS number of connections, but only if the logical parameter
'connection_filter parameter' is TRUE if FALSE it only does the KMD
filter. Figure
S1
<https://link.springer.com/article/10.1007/s13361-018-2069-9#SupplementaryMaterial>
illustrates
multiple ions associated by a repeat unit.

Finally, yes, the correct way to do it is to go through the possibilities
explored in the notebooks. See the one that fits your problem, then, the
best way would be to create a new notebook, load the packages and kendrick
mass source, as in the example notebooks and then perform the function with
your desired parameters.

Cheers,
Ricardo

Em sex, 26 de jul de 2019 às 09:22, Julie Courraud <notifications@github.com>
escreveu:

**Comment by Althalis** [comment]
Hi Ricardo, 

So I deleted lines 9-10-11 that were just a repeated example as you said.
I'll open a new issue regarding the error.
Cheers
Julie
